From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Brandon Jones <bajones@chromium.org>
Date: Tue, 15 Sep 2026 14:09:44 -0700
Subject: GL: Change mipmap workaround from flush to forcefullyFlush

An existing workaround for IMG GPUs inserts a flush prior to any
generateMipmaps calls, but a separate issue was discovered with the
same GPUs that requires a forced flush or finish to resolve. Updates
the flush to a forcefullyFlush to ensure that the issue is avoided.

Bug: chromium:549995090
Change-Id: Ifd7d45b9fcf26f4755730c770e9dae5e84688f52
Reviewed-on: https://chromium-review.googlesource.com/c/angle/angle/+/8414369
Reviewed-by: Geoff Lang <geofflang@chromium.org>
Reviewed-by: Kenneth Russell <kbr@chromium.org>
Auto-Submit: Brandon Jones <bajones@chromium.org>
Commit-Queue: Shahbaz Youssefi <syoussefi@chromium.org>

diff --git a/include/platform/gl_features.json b/include/platform/gl_features.json
index d5b71d0a69f36b5a3f010b6caadd67a6bb7d038b..a0c72fedc3f9bdc4da54fe0ecf19b4ced8481c1b 100644
--- a/include/platform/gl_features.json
+++ b/include/platform/gl_features.json
@@ -1011,7 +1011,7 @@
             "description": [
                 "Force a flush before calling generateMipmap to work around driver bugs."
             ],
-            "issue": "http://crbug.com/549350408"
+            "issue": "http://crbug.com/549350408 https://crbug.com/549995090"
         },
         {
             "name": "split_level_0_pbo_full_sub_image_2D",
diff --git a/src/libANGLE/renderer/gl/TextureGL.cpp b/src/libANGLE/renderer/gl/TextureGL.cpp
index 7602094a4eec283a0c09fca5468dc97a75f66421..c2611614f5ad57def51a546e8473f366c18e6899 100644
--- a/src/libANGLE/renderer/gl/TextureGL.cpp
+++ b/src/libANGLE/renderer/gl/TextureGL.cpp
@@ -1848,9 +1848,8 @@ angle::Result TextureGL::generateMipmap(const gl::Context *context)
 
     if (features.flushBeforeGenerateMipmap.enabled)
     {
-        // Force a flush before generating the mipmap, which avoids a bad state in the IMG driver if
-        // the texture's base level is still bound to an active FBO.
-        ANGLE_GL_TRY(context, functions->flush());
+        // Force a flush before generating the mipmap, which avoids bad states in the IMG driver.
+        ANGLE_GL_TRY(context, stateManager->forcefullyFlush());
     }
 
     stateManager->bindTexture(getType(), mTextureID);
diff --git a/src/libANGLE/renderer/gl/renderergl_utils.cpp b/src/libANGLE/renderer/gl/renderergl_utils.cpp
index ef79500456522e95ffd04652474391b9dd1d7abc..1b660db0504d8b011f784e1d7e57368b76b81bf8 100644
--- a/src/libANGLE/renderer/gl/renderergl_utils.cpp
+++ b/src/libANGLE/renderer/gl/renderergl_utils.cpp
@@ -2872,8 +2872,7 @@ void InitializeFeatures(const FunctionsGL *functions, angle::FeaturesGL *feature
     // Disable EXT_clear_texture entirely on IMG as a speculative fix for driver crashes.
     ANGLE_FEATURE_CONDITION(features, disableClearTexture, IsPowerVR(vendor));
 
-    // Forces a flush before generating a mipmap, which avoids a bad state in the IMG driver if
-    // the texture's base level is still bound to an active FBO.
+    // Forces a flush before generating a mipmap, which avoids bad states in the IMG driver.
     ANGLE_FEATURE_CONDITION(features, flushBeforeGenerateMipmap, IsPowerVR(vendor));
 
     // IMG GL drivers crash while compiling shaders with more than the limit of uniform blocks.
